shared/middleware/helpers/storage_apps.rb

```diff
@@ -140,24 +140,43 @@ def get_abuse(channel_id)
 
   # Determine if the current user can view the project
   def get_sharing_disabled(channel_id, current_user_id)
-    user_storage_owner_id, _ = storage_decrypt_channel_id(channel_id)
+    user_storage_owner_id, storage_app_id = storage_decrypt_channel_id(channel_id)
     owner_id = user_storage_ids_table.where(id: user_storage_owner_id).first[:user_id]
 
     # Sharing of a project is not disabled for the project owner
     # or the teachers of the project owner
+    # or if the current user paired with the owner
     if current_user_id == owner_id
       return false
     elsif teaches_student?(owner_id, current_user_id)
       return false
+    elsif get_user_sharing_disabled(owner_id)
+      !users_paired_on_level?(storage_app_id, current_user_id, owner_id, user_storage_owner_id)
```

Comment:
nit: I'd recommend calling renaming these `owner_id` --> `owner_user_id` and `user_storage_owner_id ` --> `owner_storage_id`